app/js/reducers/fetchOrdersReducer.js

```diff
@@ -42,6 +58,16 @@ const fetchOrdersReducer = (state = initialState, action) => {
       };
     }
 
+    case 'SORT_AND_FILTER': {
+      return {
+        ...state,
+        orders: {
+          ...state.orders,
+          results: filter(state.orders.results, action.sortBy, action.value),
+        },
```

Comment:
I don't think mutating the actual result from the database is nice. We shouldn't mutate the redux state directly like this. One edge case to note is if there are other functionalities in the application making use of this result, they would start breaking as a result of this change.
I have two suggestion to improve this block of code:
- keep the result of the filter somewhere else.
- use the state of the component holding the orders-data to track the filtering

Reply:
The logic for state change in a redux-application is done in the reducers, therefore this is appropriate, and it is generally a bad practice to fetch data from the redux state and then store the data in your components state again and then manipulate data inside the component when the reducers should have acted on it in the first place.

The actual result is manipulated by the reducer but not lost, it is stored in the local storage and the coincidence that this would break the application is next to none as the homepage is currently the only component using this reducer and the result's array always has an initial state of an empty array. therefore, during implementation, any component that would feed from the results array should be made to work with at least the *initial state* and the current implementation makes sure the result array always have at least the value at the *initial-state([])*.

Reply:
I get your point at @mogoodrich, but we can't apply any filter at that point cus the `.map` is meant to display the data that is in the `results` array and the filter is applied by the user once the DOM is loaded at least once.

If you aren't comfortable with introducing localStorage and manipulating data in the reducers, I'll be moving the logic to the component state now .

Reply:
That line is necessary. with your current code, you are neglecting the fact that a user might filter from the pre-filtered results as well and not from the whole results all the time, a perfect instance of that is when there would be more than one sort query to sortBy(e.g status, dateActivated, e.t.c). my current implementation goes beyond what we have now and accomodates future use-cases.

let me give you an instance;
let's say the data returned from the db is;

`[{ id: 1, type: 'drugorder', xxxquery: 'active' }, { id: 2, type: 'drugorder', xxxquery: inactive }, { id: 3, type: 'testorder', xxxquery: 'active'}, ]`

and a user queries by type with a value of drugorder. the filtered results would be;
`[{ id: 1, type: 'drugorder', xxxquery: 'active' }, { id: 2, type: 'drugorder', xxxquery: inactive } ]`

if a user then decide to sort using the alternate query,maybe, based on what is available would want to see only drug orders that are active, your code would filter from the first original results. thus it'll produce

`[{ id: 1, type: 'drugorder', xxxquery: 'active' }, { id: 3, type: 'testorder', xxxquery: 'active'}, ]`

and that is an undesired result.
my current implementatiion would still filter from the already filtered result and produce;

`[{ id: 1, type: 'drugorder', xxxquery: 'active' } ]` .

I hope you get the point..

Reply:
Nice catch @Efosaok